src/include/migraphx/generic_float.hpp

```diff
@@ -0,0 +1,411 @@
+        return x;
+    }
+
+    static constexpr generic_float min()
+    {
+        generic_float x{};
+        x.exponent = 1;
+        x.mantissa = 0;
+        return x;
+    }
+
+    static constexpr generic_float denorm_min()
+    {
+        generic_float x{};
+        x.exponent = 0;
+        x.mantissa = 1;
+        x.sign     = 0;
+        return x;
+    }
+
+    static constexpr generic_float lowest()
+    {
+        generic_float x{};
+        x.exponent = all_ones<ExponentSize>() - 1;
+        x.mantissa = all_ones<MantissaSize>();
+        x.sign     = 1;
+        return x;
+    }
+
+    static constexpr generic_float max()
+    {
+        generic_float x{};
+        x.exponent = all_ones<ExponentSize>() - 1;
+        x.mantissa = all_ones<MantissaSize>();
+        x.sign     = 0;
+        return x;
+    }
+
+    static constexpr generic_float epsilon()
+    {
+        generic_float x{1.0};
+        x.mantissa++;
+        return generic_float{x.to_float() - 1.0f};
+    }
+// NOLINTNEXTLINE
+#define MIGRAPHX_GENERIC_FLOAT_ASSIGN_OP(op)                        \
+    constexpr generic_float& operator op(const generic_float & rhs) \
+    {                                                               \
+        float self = *this;                                         \
+        float frhs = rhs;                                           \
+        self op frhs;                                               \
+        *this = generic_float(self);                                \
+        return *this;                                               \
+    }
+    MIGRAPHX_GENERIC_FLOAT_ASSIGN_OP(*=)
+    MIGRAPHX_GENERIC_FLOAT_ASSIGN_OP(-=)
+    MIGRAPHX_GENERIC_FLOAT_ASSIGN_OP(+=)
+    MIGRAPHX_GENERIC_FLOAT_ASSIGN_OP(/=)
+// NOLINTNEXTLINE
+#define MIGRAPHX_GENERIC_FLOAT_BINARY_OP(op)                                                   \
+    friend constexpr generic_float operator op(const generic_float& x, const generic_float& y) \
+    {                                                                                          \
+        return generic_float(float(x) op float(y));                                            \
+    }
+    MIGRAPHX_GENERIC_FLOAT_BINARY_OP(*)
+    MIGRAPHX_GENERIC_FLOAT_BINARY_OP(-)
+    MIGRAPHX_GENERIC_FLOAT_BINARY_OP(+)
+    MIGRAPHX_GENERIC_FLOAT_BINARY_OP(/)
+    MIGRAPHX_GENERIC_FLOAT_BINARY_OP(<)
+    MIGRAPHX_GENERIC_FLOAT_BINARY_OP(<=)
+    MIGRAPHX_GENERIC_FLOAT_BINARY_OP(>)
+    MIGRAPHX_GENERIC_FLOAT_BINARY_OP(>=)
+
+    friend constexpr bool operator==(const generic_float& x, const generic_float& y)
+    {
+        if(not x.is_finite() or not y.is_finite())
+            return false;
+        return std::tie(x.mantissa, x.exponent, x.sign) == std::tie(y.mantissa, y.exponent, y.sign);
+    }
+
+    friend constexpr bool operator!=(const generic_float& x, const generic_float& y)
+    {
+        return not(x == y);
+    }
+
+    constexpr generic_float& operator++() noexcept
```

Comment:
For completeness, we should add post-increment as well as pre and post decrement.